include/LightGBM/utils/common.h

```diff
@@ -1082,12 +1082,18 @@ struct __StringToTHelper<T, true> {
     // Fast (common) path: For numeric inputs in RFC 7159 format:
     const bool fast_parse_succeeded = fast_double_parser::parse_number(str.c_str(), &tmp);
 
-    // Rare path: Not in RFC 7159 format. Possible "inf", "nan", etc. Fallback to standard library:
+    // Rare path: Not in RFC 7159 format. Possible "inf", "nan", etc.
     if (!fast_parse_succeeded) {
-      std::stringstream ss;
-      Common::C_stringstream(ss);
-      ss << str;
-      ss >> tmp;
+      std::string strlower(str); 
+      std::transform(strlower.begin(), strlower.end(), strlower.begin(), [](int c) -> char { return static_cast<char>(::tolower(c)); });
+      if (strlower == std::string("inf"))
+        tmp = std::numeric_limits<double>::infinity();
+      else if (strlower == std::string("-inf"))
+        tmp = -std::numeric_limits<double>::infinity();
+      else if (strlower == std::string("nan"))
```

Comment:
Important: Missing -nan handling.

Probably best to halve the string comparisons by parsing first the "-" sign.

Reply:
Ok thanks, I'll add -nan.  I don't think it's worth obfuscating this rarely executed branch with optimisations until profiling shows it is a bottleneck.